pex/pex_bootstrapper.py

```diff
@@ -92,9 +89,32 @@ def _iter_interpreters():
     )
 
 
+def _matched_interpreters_iter(interpreters_iter, constraints):
+    candidates = []
+    failures = []
+    found = False
+
+    for interpreter in interpreters_iter:
+        if isinstance(interpreter, PythonInterpreter):
+            candidates.append(interpreter)
+            if any(interpreter.identity.matches(filt) for filt in constraints):
+                TRACER.log(
+                    "Constraints on interpreters: %s, Matching Interpreter: %s"
+                    % (constraints, interpreter.binary),
+                    V=3,
+                )
+                found = True
+                yield interpreter
+        else:
+            failures.append(interpreter)
+
+    if not found:
+        raise UnsatisfiableInterpreterConstraintsError(constraints, candidates, failures)
+
+
 def _compatible_interpreters_iter(interpreters_iter, compatibility_constraints=None):
     if compatibility_constraints:
-        for interpreter in matched_interpreters_iter(interpreters_iter, compatibility_constraints):
+        for interpreter in _matched_interpreters_iter(interpreters_iter, compatibility_constraints):
             yield interpreter
     else:
         for interpreter in interpreters_iter:
```

Comment:
This branch does not handle the switch from `PythonInterpreter.iter` to `PythonInterpreter.iter_candidates` here: https://github.com/pantsbuild/pex/pull/1028/files#diff-670bc388096d1ef0c7994c4d37976392R82 so we can get error tuples mixed into the yielded interpreters when there are no compatibility constraints.

Reply:
> so we can get error tuples

This was something I was confused by when adding type hints. In addition to the type hints, we may want to rename this to `interpreter_or_failure`:

https://github.com/pantsbuild/pex/blob/aae2a6ad55e148ada43f36c1d486af9b2312af50/pex/pex_bootstrapper.py#L97-L109

Reply:
I think the problem is better fixed by restructuring for locality. The code that calls PythonInterpreter.iter_candidates should deal with the consequences of that directly.

Reply:
I did not rename since we don't use Hungaian elsewhere but this confusing bit of code where typing is path-dependent is now typed in #1046